### ISSUE TITLE
sarama leaking memory in (*TopicMetadata).decode

### ISSUE THREAD

**Issue by chandradeepak** [Shopify/sarama#931]
##### Versions

*Please specify real version numbers or git SHAs, not just "Latest" since that changes fairly regularly.*
Sarama Version: v1.12.0-2-gac03dfa , ac03dfa0c99a940a48357980a4ebfafed5f23f9a
Kafka Version: 0.10.0.1
Go Version:1.8.3

##### Configuration

What configuration values are you using for Sarama and Kafka?

default configuration . no changes here and this have been stable for a while. 

##### Logs

When filing an issue please provide logs from Sarama and Kafka if at all
possible. You can set `sarama.Logger` to a `log.Logger` to capture Sarama debug
output.

##### Problem Description
we are seeing memory leak in sarama. this is the pprof output for it .  This happened recently. the leak is happening in decoding of the metadata from kafka . 

I see there is a similar issue addressed before https://github.com/Shopify/sarama/issues/560

But i think this is a new problem . do you think restarting kafka will get rid of this issue. But still it is strange sarama is leaking memory. i don't see any weird partition problemw in kafka . Is there any thing specific i have to check to give you more details. Below i pasted the pprof logs which showed the growth and their code lines. as you can observe , decode has 1gb in use space. 
 1.17GB    144:		err = r.Topics[i].decode(pd)

Please let me know if you need any more details . 

> 680.54MB 29.10% 29.10%  1199.06MB 51.27%  github.com/Shopify/sarama.(*TopicMetadata).decode
  329.01MB 14.07% 43.17%   329.01MB 14.07%  github.com/Shopify/sarama.(*realDecoder).getInt32Array
  306.21MB 13.09% 56.26%   306.21MB 13.09%  runtime.hashGrow
  264.04MB 11.29% 67.55%   572.75MB 24.49%  runtime.mapassign
  196.51MB  8.40% 75.95%   196.51MB  8.40%  runtime.rawstringtmp
  124.50MB  5.32% 81.27%   125.50MB  5.37%  Shopify/sarama.(*client).setPartitionCache
  102.51MB  4.38% 85.66%   102.51MB  4.38%  runtime.makemap

> /github.com/Shopify/sarama/metadata_response.go
  680.54MB     1.17GB (flat, cum) 51.27% of Total
         .          .     67:	if err != nil {
         .          .     68:		return err
         .          .     69:	}
         .          .     70:	tm.Err = KError(tmp)
         .          .     71:
         .   189.51MB     72:	tm.Name, err = pd.getString()
         .          .     73:	if err != nil {
         .          .     74:		return err
         .          .     75:	}
         .          .     76:
         .          .     77:	n, err := pd.getArrayLength()
         .          .     78:	if err != nil {
         .          .     79:		return err
         .          .     80:	}
         .          .     81:	tm.Partitions = make([]*PartitionMetadata, n)
         .          .     82:	for i := 0; i < n; i++ {
  680.54MB   680.54MB     83:		tm.Partitions[i] = new(PartitionMetadata)
         .   329.01MB     84:		err = tm.Partitions[i].decode(pd)
         .          .     85:		if err != nil {
         .          .     86:			return err
         .          .     87:		}
         .          .     88:	}
         .          .     89:

> /vendor/github.com/Shopify/sarama/metadata_response.go
         0   329.01MB (flat, cum) 14.07% of Total
         .          .     23:	pm.Leader, err = pd.getInt32()
         .          .     24:	if err != nil {
         .          .     25:		return err
         .          .     26:	}
         .          .     27:
         .   166.50MB     28:	pm.Replicas, err = pd.getInt32Array()
         .          .     29:	if err != nil {
         .          .     30:		return err
         .          .     31:	}
         .          .     32:
         .   162.50MB     33:	pm.Isr, err = pd.getInt32Array()
         .          .     34:	if err != nil {
         .          .     35:		return err
         .          .     36:	}
         .          .     37:
         .          .     38:	return nil

> /vendor/github.com/Shopify/sarama/metadata_response.go
       4MB     1.18GB (flat, cum) 51.46% of Total
         .          .    122:	n, err := pd.getArrayLength()
         .          .    123:	if err != nil {
         .          .    124:		return err
         .          .    125:	}
         .          .    126:
         .          .    127:	r.Brokers = make([]*Broker, n)
         .          .    128:	for i := 0; i < n; i++ {
       2MB        2MB    129:		r.Brokers[i] = new(Broker)
         .   512.04kB    130:		err = r.Brokers[i].decode(pd)
         .          .    131:		if err != nil {
         .          .    132:			return err
         .          .    133:		}
         .          .    134:	}
         .          .    135:
         .          .    136:	n, err = pd.getArrayLength()
         .          .    137:	if err != nil {
         .          .    138:		return err
         .          .    139:	}
         .          .    140:
         .          .    141:	r.Topics = make([]*TopicMetadata, n)
         .          .    142:	for i := 0; i < n; i++ {
       2MB        2MB    143:		r.Topics[i] = new(TopicMetadata)
         .     1.17GB    144:		err = r.Topics[i].decode(pd)
         .          .    145:		if err != nil {
         .          .    146:			return err
         .          .    147:		}
         .          .    148:	}
         .          .    149:

**Comment by eapache** [comment]
Are you producing or consuming? Can you provide some Sarama logs?

Go is a garbage-collected language, so the only way to leak memory is to hold a reference to something accidentally. Are you holding a reference to all consumed messages or something like that?

**Comment by chandradeepak** [comment]
@eapache , Hi i am doing both producing and consuming. This happens especially in consumers.  We are not holding any memory to produced or consumed message . this was all working well until recently . As you can see the metadata was allocating too much memory and that keeps increasing. That is the cause of the leak. we have around 2000 topics and 5 partitions each so that should not be much to allocate a 1GB memory right .

**Comment by eapache** [comment]
> working well until recently

What changed? Did you upgrade Sarama or Kafka? Adjust some configuration? Add partitions?

> As you can see the metadata was allocating too much memory and that keeps increasing. That is the cause of the leak. we have around 2000 topics and 5 partitions each so that should not be much to allocate a 1GB memory right .

2k topics with 5 partitions each is 10k partitions. Each partition takes maybe a few-hundred bytes to store (call it 1KB to be safe) in which case the entire set would take around 10MB. Not small, but nowhere near 1GB I wouldn't think.

I have no idea what could be causing this. As I mentioned, Go has a garbage-collector, so there are really only two possible causes for this:
- Sarama is legitimately trying to allocate this much memory. The math just doesn't add up unless my estimates are way off. How many replicas do you maintain of each partition?
- Something is holding on to stale metadata responses. I don't know of anything in Sarama that would hold onto this data, and it's not exposed to the user much so again I'm not sure how this could happen.

**Comment by chandradeepak** [comment]
we haven't until any thing recently. we restarted some of our clients and saw this happen.  and we have 3 replicas for each partition . we have never seen this much allocation in metadata decode previously , as you can see in pprof clearly so much is being allocated in metadata decode and that bloats the client.  even though go is garbage collected if the allocation happens too quick that process uses too much memory right . that is what happening in our case. 

what is stale metadata response ? 

Assuming restarting kafka will fix this problem we did that and we saw these errors in kafka logs ,

we have seen kafka errors with   Error for partition [topicA,3] to broker 1001:org.apache.kafka.common.errors.NotLeaderForPartitionException: This server is not the leader for that topic-partition. (kafka.server.ReplicaFetcherThread)

For most of the topics.

**Comment by eapache** [comment]
> even though go is garbage collected if the allocation happens too quick that process uses too much memory right . that is what happening in our case.

Right, so then it's not exactly a leak, since the memory isn't gone forever. It's still a major problem though.

> what is stale metadata response ?

I am wondering if something was causing Sarama to hold onto old metadata, not just the current metadata, but I don't know what that could be.

> org.apache.kafka.common.errors.NotLeaderForPartitionException: This server is not the leader for that topic-partition. (kafka.server.ReplicaFetcherThread)

This is a pretty normal error to happen temporarily when you restart a broker, since leadership for its partitions will move around as it restarts.

Do you have any logs from Sarama around the time that memory started increasing?

The only thing I can think of is that the kafka brokers started returning incorrect metadata which is causing Sarama to ask for it again and again, and since you have so many topics the garbage collector can't keep up with the amount of metadata that Sarama is parsing? But really I'm just guessing.

**Comment by chandradeepak** [comment]
Hello @eapache , i think i know what is happening on the leak part. Before that i followed your the recommendation of creating each client per producer and consumer as per your recommendation. 

"Client is a generic Kafka client. It manages connections to one or more Kafka brokers. You MUST call Close() on a client to avoid leaks, it will not be garbage-collected automatically when it passes out of scope. It is safe to share a client amongst many users, however Kafka will process requests from a single client strictly in serial, so it is generally more efficient to use the default one client per producer/consumer."

we have 3000 topics . and 5 partitions per topic. so it is 15000 partitions at any time.  previously we were reusing the clients across different publisher. But according to this recommendation i made publisher to client one to one . That is i call NewAsynProducer instead of NewAsyncProducerFromClient. 

 coming to the leak assuming the data allocated during that time for creation PartitionMetaData is 
680.54MB 680.54MB 83:	tm.Partitions[i] = new(PartitionMetadata)
considering each PartitionMetadata structure is 80 bytes . I got 15000 partitions . To find out how many clients are created it would be some thing like ( 680 * 1024 * 1024)/ (3000 * 5 * 80 )  = 600. So around 600 connections which is close to what we have . 

so assuming that if number of connections increase it memory usage will increases correct  ?

So one improvement that can be done is the 
func (client *client) backgroundMetadataUpdater()  where we call client.RefreshMetadata() with no topics passed. So it gets the partitionMetaData for all the topics even thought if that NewAsynProducer is used for ingesting data for one topic only . which is overkill . And if we have close to 3000 topics basically we are allocating data for all those asyncproducer unnecessarily. 

so do you think a config variable has to be exposed to restrict a client for ingesting one topic has to get metadata for that topic alone or it is better to reuse the existing clients using the NewAsyncProducerFromClient.  please suggest which is better approach.  

But in general client getting metadata for all topics irrespective of what it is being used for has to be fixed. since there are other parts of the code where individual topic metadata is fetched when required. 

your thoughts ....

**Comment by eapache** [comment]
It sounds like you are creating a new asyncproducer for every topic? This is very inefficient and unnecessary, you can use one asyncproducer for many many topics.

**Comment by chandradeepak** [comment]
@eapache but one client ingest data to one topic only. so you suggest reusing the asyncproducer across client ?

**Comment by eapache** [comment]
> but one client ingest data to one topic only. so you suggest reusing the asyncproducer across client ?

Do you mean client as `sarama.Client` or client as an entire go program? A single `sarama.Client` can handle as many topics as you want. You can use `sarama.AsyncProducer` to produce to multiple topics, and you can use `sarama.Consumer` to consume from multiple topics.

**Comment by chandradeepak** [comment]
This pull request seems to be working . I also started reusing the async producer for different clients.  It seems stable now. will let you know if any thing comes up. But thanks for the quick fix.

**Comment by chandradeepak** [comment]
Hello @eapache

The issue is not fixed yet.  I ran it on weekend and i see the memory increased again and crashed our systems. One new thing i found it happens only in the subscriber part not on publisher part. **we create separate clients for publishers and subscribers**.  So for publishers i see that only metadata for the topic alone is being retrieved.

For producer the metadata is always updated to have one topic only . But for all the clients created 
for subscribers the metadata is never retrieved for one topic or couple of topics for which we need data for. so the intial value with all topics is still being passed and it retrieves the metadata for all topics again . 

do we think we can update that ?  how can i reopen the issue ?

**Comment by eapache** [comment]
> the metadata is never retrieved for one topic or couple of topics for which we need data for. so the intial value with all topics is still being passed and it retrieves the metadata for all topics again.

I find this very surprising, what do you have `Config.Metadata.RefreshFrequency` set to? Once you call `ConsumePartition` on the consumer it will fetch that one topic which should happen well before the first background refresh.

**Comment by chandradeepak** [comment]
my Config.Metadata.RefreshFrequency=10min . It is the default value.  I looked at the pull request . I found the issue is in RefreshMetaData. so fixing it in backgroundMetadataUpdater might not solve the issue.  In the code of sarama-cluster , in the file https://github.com/bsm/sarama-cluster/blob/master/consumer.go


func (c *Consumer) refreshMetadata() error {
	err := c.client.RefreshMetadata()
	if err == sarama.ErrTopicAuthorizationFailed {
		// maybe we didn't have authorization to describe all topics
		err = c.client.RefreshMetadata(c.coreTopics...)
	}
	return err
}

So for any one calling RefreshMetadata() this would be an empty call and that would get all the topics. Now next time the backgroundMetadataUpdater comes it gets its list of topics from metadata which is already filled. (client.Topics() , which gets its data from a map which is client.metadata which is already filled when the first call was made.)


	if !client.conf.Metadata.Full {
				if specificTopics, err := client.Topics(); err != nil {
					Logger.Println("Client background metadata topic load:", err)
					break
				} else {
					topics = specificTopics
				}
			}

This is because in refreshMetadata we don't start from scratch where we delete all the topics and partitions in metadata and start over instead we rely on metadata already present .  so i had to restrict from getting all the metadata topics first time to stop the growth .
